### DISCUSSION TITLE
Breaking changes made on sd-webui-controlnet

### DISCUSSION THREAD

**Discussion by huchenlei** [lllyasviel/stable-diffusion-webui-forge#40] [General]
Here is a list of changes made to sd-webui-controlnet extension, that is not backward compatible.

## Changes
- Low VRAM checkbox is removed, as forge has a global memory management system.
- Hr Option is temporarily removed.
- Crop input image based on A1111 input checkbox removed. IPAdapter and many other control types now do not crop input image by default.
- Multi-inputs is removed. You now need to enable multiple IP-Adapter units to achieve effect of previous multi-inputs. We will add this back if there is a strong demand on that. The new batch upload tab is not equivalent to previous multi-inputs tab.
- Many preprocessors are renamed.
- Batch loopback checkbox is removed.

## Update 2024-02-07
Hr Option is added back. https://github.com/lllyasviel/stable-diffusion-webui-forge/pull/90

## Update 2024-02-09
ControlNet models choice is filtered by active SD model's version. https://github.com/lllyasviel/stable-diffusion-webui-forge/pull/131

## Update 2024-02-10
- ControlNet API supported added back. https://github.com/lllyasviel/stable-diffusion-webui-forge/pull/162
- `save_detected_map` API param added back.

Note:
Now the only way to pass image/mask to ControlNet unit is
```python
{
  "image": $base64string,
  "image_mask": $base64string,
  ...
}
```
All other syntax are no longer supported. You might need to adjust your script accordingly.

**Comment by strawberrymelonpanda** [comment]
> Multi-inputs is removed. You now need to enable multiple IP-Adapter units to achieve effect of previous multi-inputs
> We will add this back if there is a strong demand on that.

Yes, please!  I was so stoked to see it added, and bummed to see it's now gone again.  It's super important for IP Adapter "Instant LoRA" type workflows, and a whole lot faster than adding 2-4 different duplicate ControlNets to do the same thing.  (Edit: #98)

> Crop input image based on A1111 input checkbox removed. IPAdapter and many other control types now do not crop input image by default.

I haven't had a chance to try Forge yet, I will soon, but this one is a little concerning.  I had horrible results without that checkbox, which were fixed entirely when it was added (and a few fixes came in).

**Comment by Evilinside3357** [comment]
>     * Batch loopback checkbox is removed.

Please bring back this option. It's such a crucial element in Vid2Vid workflow when working frame by frame

**Comment by CRCODE22** [comment]
The removal of multi inputs is a deal breaker for me and on top of that not allowing to use the main branch of the control net. I am going back to automatic 1111 where I can use multi inputs.

**Comment by Moussack** [comment]
Why removed multi input though?? it's not breaking anything or what?

my workaround right now is using faceswaplab to blend the faces and use that blended face image to the controlnet.
please bring multi input back thx.

**Comment by JaB43a** [comment]
With the multiple input option you get more precise facial refinement/recognition. Please, another request here so that it is present again in WEBUI

**Comment by lhucklen** [comment]
Please bring it back.

**Comment by sabum6800** [comment]
-- Multi - Inputs --

I never wrote in a Forum ever! But this is important and you have written: 

> Multi-inputs is removed. You now need to enable multiple IP-Adapter units to achieve effect of previous multi-inputs. We will add this back if there is a strong demand on that. The new batch upload tab is not equivalent to previous multi-inputs tab.

So have my DEMAND here! 💯

**Comment by brianknight144** [comment]
You have my DEMAND too. I don't know why this option is removed...

**Comment by gohan2091** [comment]
Why is Multi input removed? this is really really important to me. Please add it back

**Comment by DA-Charlie** [comment]
Crazy how it's been months now we're begging for Batch loopback.
